pkg/render/gateway_api.go

```diff
@@ -0,0 +1,507 @@
+
+	// Substitute possibly modified image names into the envoy-gateway-config.
+	envoyGatewayConfig := resources.envoyGatewayConfig.DeepCopyObject().(*envoyapi.EnvoyGateway)
+	envoyGatewayConfig.Provider.Kubernetes.RateLimitDeployment.Container.Image = &pr.envoyRatelimitImage
+	envoyGatewayConfig.Provider.Kubernetes.ShutdownManager.Image = &pr.envoyGatewayImage
+
+	// Add pull secrets into the envoy-gateway-config, insofar as the API allows this.  (In
+	// other words, currently only for the ratelimit image.  So effectively the gateway image
+	// has to be pullable without any pull secrets.)
+	if envoyGatewayConfig.Provider.Kubernetes.RateLimitDeployment.Pod == nil {
+		envoyGatewayConfig.Provider.Kubernetes.RateLimitDeployment.Pod = &envoyapi.KubernetesPodSpec{}
+	}
+	envoyGatewayConfig.Provider.Kubernetes.RateLimitDeployment.Pod.ImagePullSecrets = secret.GetReferenceList(pr.cfg.PullSecrets)
+
+	// Rebuild the ConfigMap with those changes.
+	envoyGatewayConfigMap := resources.envoyGatewayConfigMap.DeepCopyObject().(*corev1.ConfigMap)
+	if bytes, err := yaml.Marshal(*envoyGatewayConfig); err == nil {
+		envoyGatewayConfigMap.Data[EnvoyGatewayConfigKey] = string(bytes)
+	} else {
+		panic(fmt.Sprintf("couldn't marshal EnvoyGateway to YAML: %v", err))
+	}
+
+	objs = append(objs, envoyGatewayConfigMap)
+
+	// Deep copy the controller deployment,
+	controllerDeployment := resources.controllerDeployment.DeepCopyObject().(*appsv1.Deployment)
+
+	// Substitute possibly modified gateway image name.
+	controllerDeployment.Spec.Template.Spec.Containers[0].Image = pr.envoyGatewayImage
+
+	// Reference additional pull secrets.
+	controllerDeployment.Spec.Template.Spec.ImagePullSecrets = append(
+		controllerDeployment.Spec.Template.Spec.ImagePullSecrets,
+		secret.GetReferenceList(pr.cfg.PullSecrets)...)
+
+	// Add a k8s-app label that we can use to provide API access for the controller.
+	controllerDeployment.Spec.Template.Labels["k8s-app"] = GatewayControllerLabel
+
+	// Apply customizations from the GatewayControllerDeployment field of the GatewayAPI CR.
+	rcomp.ApplyDeploymentOverrides(controllerDeployment, pr.cfg.GatewayAPI.Spec.GatewayControllerDeployment)
+
+	objs = append(objs, controllerDeployment)
+
+	// Deep copy the certgen job,
+	certgenJob := resources.certgenJob.DeepCopyObject().(*batchv1.Job)
+
+	// Substitute possibly modified gateway image name.
+	certgenJob.Spec.Template.Spec.Containers[0].Image = pr.envoyGatewayImage
+
+	// Reference additional pull secrets.
+	certgenJob.Spec.Template.Spec.ImagePullSecrets = append(
+		certgenJob.Spec.Template.Spec.ImagePullSecrets,
+		secret.GetReferenceList(pr.cfg.PullSecrets)...)
+
+	// Apply customizations from the GatewayCertgenJob field of the GatewayAPI CR.
+	rcomp.ApplyJobOverrides(certgenJob, pr.cfg.GatewayAPI.Spec.GatewayCertgenJob)
+
+	objs = append(objs, certgenJob)
+
+	objs = append(objs, secret.ToRuntimeObjects(secret.CopyToNamespace(resources.namespace.Name, pr.cfg.PullSecrets...)...)...)
```

Comment:
I was wondering if it would be better to render the pull secrets before you render the deployments to avoid k8s v1/Events, like `FailedToRetrieveImagePullSecret`.

Reply:
Thanks.  I wasn't sure if I could fully rely on the object ordering, and hence if the ordering really mattered.  But it makes sense that it would do.